### PR TITLE
K epsilon theta^2, new PBL closure, Zonato et al, 2022

### PR DESCRIPTION
**Pull request by andreazonato**
Incorporation of a k-epsilon-theta^2 PBL scheme, based on [Zonato et al, 2022](https://journals.ametsoc.org/view/journals/mwre/150/8/MWR-D-21-0299.1.xml)
TYPE: new feature

KEYWORDS: k-epsilon, KEPS, TKE,PBL, TPE

SOURCE: Andrea Zonato, Department of Environmental, Civil and Mechanical Engineering, University of Trento, Trento, Italy

DESCRIPTION OF CHANGES:
1. The file module_bl_keps.F includes the module for this new scheme.
2.  modification in addtenc,pbl_driver
3. 5 new variables have been added:
diss_pbl (take dissipation rate, state)
tpe_pbl (temperature variance, state)
diss_adv(take dissipation rate advection, scalar)
tpe_adv(temperature variance advection, scalar)
the file module_first_rk_step_part1.F and Registry.EM_COMMON are revised accordingly, and it is defined as kepsscheme and bl_pbl_physics=17

LIST OF MODIFIED FILES:
phys/module_bl_keps.F
phys/module_pbl_driver.F
phys/module_physics_addtendc.F
phys/Makefile
Registry/Registry.EM_COMMON
dyn_em/module_first_rk_step_part1.F

TESTS CONDUCTED: 
1. Do mods fix problem? How can that be demonstrated, and was that test conducted?
2. The Jenkins tests have passed.

RELEASE NOTE: A new k-epsilon-theta^2 PBL scheme, including two additional prognostic equations for dissipation rate and temperature variance, has been included in the WRF model ([Zonato et al., 2022](https://journals.ametsoc.org/view/journals/mwre/150/8/MWR-D-21-0299.1.xml) )

### CODE REVIEW

**Comment by dudhia** [comment]
Hi Andrea, also remember to send a letter of intent to the WRF Physics Review Panel.
https://www2.mmm.ucar.edu/wrf/users/code_contribution/physics_review.php

**Comment by dudhia** [comment]
@weiwangncar could you find which checks failed? Thanks.

**Comment by weiwangncar** [comment]
@andreazonato @dudhia There are a couple of things that failed:
1. changes in file module_first_rk_step_part1.F are not added in this PR. This basically caused every compilation to fail.
2. newer variables may need to be declared in Registry.wrfvar, Registry.tladj to enable WRFDA and WRF-Plus to compile.

**Comment by andreazonato** [comment]
Thank you. I added the module_first_rk_step_part1.F. However, I guess additional variables should not be added in the registries you mentioned.

**Comment by weiwangncar** [comment]
It looks like the reg tests have passed:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           58
    Number of Simulations  : 158           156        0
    Number of Comparisons  : 95           92        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Comment by weiwangncar** [comment]
@andreazonato In [PR#1303](https://github.com/wrf-model/WRF/pull/1303), a TKE scheme with predictive dissipation rate (epsilon) was added to WRF. I am wondering if there are variables your new scheme can share with that scheme?

**Comment by andreazonato** [comment]
Hi Wei,

actually yes. The variables that can be shared are:
state   real   pep_pbl        ikj    misc        1         -     hr        "PEP_PBL"            "TKE dissipation rate from EEPS"    "m2 s-3"
state   real   pek_adv        ikjftb  scalar      1         -      i0rusdf=(bdy_interp:dt) "pek_adv"       "TKE from EEPS"      "m2 s-2"
state   real   pep_adv        ikjftb  scalar      1         -      i0rusdf=(bdy_interp:dt) "pep_adv"       "TKE dissipating rate from EEPS"      "m2 s-3"

however, I called them in another way, namely:

diss_pbl,tke_adv and diss_adv.

But they can be changed. Moreover, there are 2 variables that are brand new:

 tpe_pbl ,tpe_adv (theta^2 and theta^2 advection)

Andrea

**Comment by dudhia** [comment]
I prefer the names Andrea has.

File: phys/module_pbl_driver.F [hunk 1]
```diff
@@ -193,6 +195,8 @@ SUBROUTINE pbl_driver(                                          &
 #if (EM_CORE==1)
    USE module_bl_mynn
    USE module_bl_eeps
+   USE module_bl_keps
+   USE module_bl_keps
```
Comment (dudhia) [inline phys/module_pbl_driver.F:199]: seems to be a repeated line here

**Review by dudhia — APPROVED** [review]
Regarding comments, we can leave the names of variables as they are, in my opinion.

File: Registry/Registry.EM_COMMON [hunk 2]
```diff
@@ -3115,6 +3121,7 @@ package   temfpblscheme  bl_pbl_physics==10          -             state:te_temf
 package   shinhongscheme bl_pbl_physics==11          -             state:el_pbl,tke_pbl
 package   gbmpblscheme   bl_pbl_physics==12          -             state:exch_tke,el_pbl,tke_pbl
 package   eepsscheme     bl_pbl_physics==16          -             scalar:pek_adv,pep_adv;state:tke_pbl,pep_pbl
+package   kepsscheme     bl_pbl_physics==17          -             scalar:tke_adv,diss_adv,tpe_adv;state:tpe_pbl,diss_pbl,tke_pbl,wu_tur,wv_tur,wt_tur,wq_tur
 package   mrfscheme      bl_pbl_physics==99          -             -
```
Comment (weiwangncar) [inline Registry/Registry.EM_COMMON:3125]: @andreazonato Need to add PR_PBL to the package.

File: dyn_em/module_first_rk_step_part1.F [hunk 3]
```diff
@@ -1154,8 +1154,11 @@ SUBROUTINE first_rk_step_part1 (   grid , config_flags              &
      &          ,SF_URBAN_PHYSICS=config_flags%sf_urban_physics                          &
 ! Bep changes end
 ! add tke_pbl, and turbulent fluxes
-     &          ,TKE_PBL=grid%tke_pbl,EL_PBL=grid%el_pbl,WU_TUR=grid%wu_tur              &
-     &          ,WV_tur=grid%wv_tur,WT_tur=grid%wt_tur,WQ_tur=grid%wq_tur                &
+     &          ,TKE_PBL=grid%tke_pbl,DISS_PBL=grid%diss_pbl,TPE_PBL=grid%tpe_pbl       &
+     &          ,TKE_ADV=scalar(ims,kms,jms,P_tke_adv),DISS_ADV=scalar(ims,kms,jms,P_diss_adv) &
+     &          ,TPE_ADV=scalar(ims,kms,jms,P_tpe_adv),PR_PBL=grid%pr_pbl  &
+     &          ,EL_PBL=grid%el_pbl,WU_TUR=grid%wu_tur              &
+     &          ,WV_tur=grid%wv_tur,WT_tur=grid%wt_tur,WQ_tur=grid%wq_tur
```
Comment (weiwangncar) [inline dyn_em/module_first_rk_step_part1.F:1161]: @andreazonato Need a '&' at the end of the line. Can also line the lines up better.

File: phys/module_pbl_driver.F [hunk 4]
```diff
@@ -537,7 +540,10 @@ SUBROUTINE pbl_driver(                                          &
                                                          RVBLTEN, &
                                                         RTHBLTEN, &
                                            EXCH_H,EXCH_M,TKE_PBL, &
-                                                        RTHRATEN  ! for GBM PBL scheme
+                                                DISS_PBL,TPE_PBL, &
+                                        TKE_ADV,DISS_ADV,TPE_ADV, &
+                                                 PR_PBL,RTHRATEN  ! for GBM PBL scheme
```
Comment (weiwangncar) [inline phys/module_pbl_driver.F:545]: @andreazonato If you add PR_PBL on this line, then the comment at the end isn't correct any more. Can remove the comment.

Comment (andreazonato) [reply phys/module_pbl_driver.F:545]: Yes, all modifications make sense. Shall I do it or you already made them?

Comment (dudhia) [reply phys/module_pbl_driver.F:545]: @andreazonato you can do it

**Comment by weiwangncar** [comment]
@andreazonato Please go ahead and add the changes that you agree. Thanks.

**Comment by weiwangncar** [comment]
@andreazonato I've made the changes, and hope they are ok with you.

**Comment by weiwangncar** [comment]
@dudhia Need you to approve this PR again.

**Comment by weiwangncar** [comment]
The last Jenkins test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Comment by andreazonato** [comment]
sorry @weiwangncar, I was traveling overseas. I checked the changes and they work for me.

Thank you.

**Comment by weiwangncar** [comment]
@dudhia Please re-approve this PR.

**Review by weiwangncar — APPROVED** [review]

**Comment by andreazonato** [comment]
first_test
second_test

@weiwangncar 
Both tests passed successfully

**Review by dudhia — APPROVED** [review]
Regarding comments, we can leave the names of variables as they are, in my opinion.

**Comment by weiwangncar** [comment]
@andreazonato Can you comment on which surface layer schemes work with this PBL?

**Comment by andreazonato** [comment]
I've actually no idea about the compatibility with the surface schemes.
I tested it with sf_sfclay_physic=1, and the variables I take in input are the same of BouLac.

So I guess it work at least with module_sf_sfclayrev and modue_sf_sfclay

**Comment by weiwangncar** [comment]
@andreazonato I did a test over East Asia with a 15 km grid (the domain covers Tibet). When using YSU scheme, a time step of 90 sec can be used to stably integrate the model. When switching to your scheme, even a 60 sec time step will not finish a 24-h run (model stopped after 4+ hours). This is with surface layer physics option 1. The case is in May. Using sfclay=91, the model stopped after 2:30 forecast, and sfclay=2, 8:45 min. If you have any ideas, let us know.

**Comment by dudhia** [comment]
Maybe try a finer grid that it was designed for.

**Comment by weiwangncar** [comment]
@andreazonato Those tests were done with diff_opt=2, km_opt=4. With diff_opt=1, the outcome improves. Using sfclay option 2, the test can complete with 90 sec time step. With sfclay option 1, the test can complete with 60 sec time step, but fails with time steps of 75 sec.

**Comment by andreazonato** [comment]
Thank you @weiwangncar  for the tests.

Can you please let me know where the failings take place? Could I have a look at the simulations?
I had some issues at the first atmospheric level with the equation of temperature variance. That may be the issue.


Andrea

**Comment by weiwangncar** [comment]
@andreazonato Thanks for your comment. The model actually died not in the PBL scheme, but in other part of the code. For example, when surface layer scheme option 1 is used, the model would die inside surface layer scheme. When using sfclay option 91 and 2, the code died in RRTMG_LW. So we need to track the origin of the problem. Do you have access to Cheyenne? If so, I will let you know where to find my test data.

**Comment by andreazonato** [comment]
Yes, I do have access to cheyenne. Let me know where I can access the simulations.
There may be some issue in the first level of the atmosphere, as I was experiencing.

**Comment by weiwangncar** [comment]
@andreazonato I sent you the info to your gmail account.

**Pull request merged**